engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionCallbacks.java

```diff
@@ -214,19 +216,32 @@ private FunctionCallInfo functionCallInfoById(UUID nodeId) {
     return calls.get(nodeId);
   }
 
-  private String[] typeOf(Object value) {
+  private TypeInfo typeOf(Object value) {
     if (value instanceof UnresolvedSymbol) {
-      return new String[] {Constants.UNRESOLVED_SYMBOL};
+      return TypeInfo.ofType(Constants.UNRESOLVED_SYMBOL);
     }
 
-    var typeOfNode = TypeOfNode.getUncached();
-    Type[] allTypes = value == null ? null : typeOfNode.findAllTypesOrNull(value, true);
-    if (allTypes != null) {
-      String[] result = new String[allTypes.length];
-      for (var i = 0; i < allTypes.length; i++) {
-        result[i] = getTypeQualifiedName(allTypes[i]);
+    if (value != null) {
+      final TypeOfNode typeOfNode = TypeOfNode.getUncached();
+      final Type[] publicTypes = typeOfNode.findAllTypesOrNull(value, false);
+
+      if (publicTypes != null) {
+        final Type[] allTypes = typeOfNode.findAllTypesOrNull(value, true);
+        // Relies on the fact that allTypes appends extra types to the end of publicTypes.
```

Comment:
Make this safer:
- assert the first `publicType.length` part is the same in both arrays
- change the format of the message